[WIP] Element RequestedTheme, ActualTheme

MartinZikmund opened the pull request:
GitHub Issue (If applicable): closes #3302 

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

Element-level theming is not supported.


## What is the new behavior?




## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

nventive-devops commented on src/Uno.UI/UI/Xaml/IFrameworkElement.cs:
```diff
@@ -63,6 +63,7 @@ public partial interface IFrameworkElement : IUIElement, IDataContextProvider, I
 		event RoutedEventHandler Unloaded;
 		event EventHandler<object> LayoutUpdated;
 		event SizeChangedEventHandler SizeChanged;
+		//event TypedEventHandler<FrameworkElement, object> ActualThemeChanged;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5709307)

nventive-devops commented on src/Uno.UI/UI/Xaml/IFrameworkElement.cs:
```diff
@@ -98,6 +99,10 @@ public partial interface IFrameworkElement : IUIElement, IDataContextProvider, I
 
 		Windows.UI.Xaml.Media.Brush Background { get; set; }
 
+		//ElementTheme RequestedTheme { get; set; }
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5709307)

jeromelaban commented on src/Uno.UI/UI/Xaml/FrameworkElement.cs:
```diff
@@ -357,6 +365,90 @@ private void UpdateActiveStyle()
+			DependencyProperty.Register(
+				nameof(ActualTheme), typeof(ElementTheme),
+				typeof(FrameworkElement),
+				new PropertyMetadata(ElementTheme.Default, ActualThemePropertyChanged));
```
Should this property be property inherited ?

MartinZikmund replied:
Adjusted :-)

dr1rrb commented on src/Uno.UI/UI/Xaml/FrameworkElement.cs:
```diff
@@ -357,6 +365,90 @@ private void UpdateActiveStyle()
+
+		private static void ActualThemePropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
+		{
+			if (args.NewPrecedence < DependencyPropertyValuePrecedences.DefaultValue)
```
That's smart! Like it!

dr1rrb commented on src/Uno.UI/UI/Xaml/ResourceDictionary.cs:
```diff
@@ -159,12 +159,12 @@ private void TryMaterializeLazy(object key, ref object value)
 		}
 
-		private bool GetFromMerged(object key, out object value)
+		private bool GetFromMerged(object key, out object value, ElementTheme elementTheme = ElementTheme.Default)
```
I would assume that this parameter should not be optional, or at least, it should default to the `RequestedTheme` (same thing for all other methods that now accepts a theme as parameter).

davidjohnoliver replied:
+1, I prefer not to have default parameter values on private or internal methods, unless it'd really introduce a lot of noise otherwise.

davidjohnoliver commented on src/Uno.UI/UI/Xaml/FrameworkElement.cs:
```diff
@@ -357,6 +365,90 @@ private void UpdateActiveStyle()
+				typeof(FrameworkElement),
+				new FrameworkPropertyMetadata(ElementTheme.Default, FrameworkPropertyMetadataOptions.Inherits, RequestedThemePropertyChanged));
+
+		private static void RequestedThemePropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
```
Should be `OnRequestedThemePropertyChanged` per normal convention

davidjohnoliver commented on src/Uno.UI/UI/Xaml/FrameworkElement.cs:
```diff
@@ -357,6 +365,90 @@ private void UpdateActiveStyle()
+					}
+				}
+			}
+			PropagateThemeChanged(frameworkElement);			
```
I'm not sure why you need this - isn't the point of making the DP inherited that you _don't_ have to manually propagate the change?

MartinZikmund replied:
Adjusted :-)

davidjohnoliver commented on src/Uno.UI/UI/Xaml/FrameworkElement.cs:
```diff
@@ -357,6 +365,90 @@ private void UpdateActiveStyle()
+			}
+		}
+
+		private static void ActualThemePropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
```
should be `OnActualThemePropertyChanged`

davidjohnoliver commented on src/Uno.UI/UI/Xaml/ResourceDictionary.cs:
```diff
@@ -188,7 +188,15 @@ private bool ContainsKeyMerged(object key)
+			var theme = Themes.Active;
+			if (elementTheme != ElementTheme.Default)
+			{
+				theme = elementTheme.ToString("g");
```
Not sure if calling enum.ToString() has potential perf issues - a switch block might be safer.

MartinZikmund replied:
Adjusted :-)

MartinZikmund commented:
@davidjohnoliver Have you been able to get to #3388 yet, please? I have noticed it behaves unusually even for some non-visual state items. I will update the issue with a comment :-) .

MartinZikmund committed:
feat(Themes): ElementTheme, ActualTheme

MartinZikmund commented:
@davidjohnoliver @jeromelaban @dr1rrb Pushed new version, made several changes to the logic

- both dependency properties now use inheritance
- theme dictionary search does fallback to `Default` if the requested theme dictionary exists, but the searched key does not
- added element theme support to some code paths which I previously didn't notice
- removed default parameter values to make `ElementTheme` explicit

There is still a bit more work required, as I noticed `BasicThemeResources` sample does not yet work the same way as in UWP, so I will continue with this

MartinZikmund commented:
@davidjohnoliver Maybe I found why the `TextBlock` color didn't update in my Messages sample - it seems that when `TextBlock` color is not explicitly set, it does not inherit its value from the container control's foreground. This causes an issue as the color stays on its default value. You can see this currently on the `BasicThemeResources` page when you switch to **Local Dark** theme - this top `TextBlock` which stays always black:
![image](https://user-images.githubusercontent.com/1075116/86453231-bc5af400-bd1d-11ea-9ff5-f9a0c1bd79be.png)

davidjohnoliver commented:
> @davidjohnoliver
> 
> ## Update
> Maybe this will be a bit more complicated :-D . Consider the following:

Indeed it may not be straightforward :D It seems to touch on this issue - https://github.com/unoplatform/uno/issues/3443 . Inheritance predecence is tricky.

MartinZikmund committed:
chore: Improving RequestedTheme support

MartinZikmund committed:
test(Themes): Add tests

MartinZikmund committed:
chore: Undo inherit RequestedTheme, use precedence for ResourceBinding

MartinZikmund commented:
@davidjohnoliver Made another round of significant changes:

- `RequestedTheme` no longer inherits - I noticed that setting it on a scope in UWP does not affect the children. Hence now changing performs a "tunneling" similar to when Application theme is set - both are now refactored to use a common `FrameworkElement` logic
- I had to add precedence to `ResourceBinding`, as default style theme resources were overwriting the explicit style resources (as theme resource setting had previously used local precedence). I hope I didn't miss something, please double check this portion especially.
- I had to add logic to update theme resources when parent of an element is changed. This is needed because during parsing, the parent is not yet set, so its resources cannot be found.
- I have updated the `ResourceDictionary` indexer to throw when key is not found, to match UWP. I have found a commented out test for this, however so I am not sure if this was done for some specific reason? Maybe because we expect some resources may not yet exist like reveal brushes?

What is missing is to adjust the `Foreground` property which is super tricky. The current inheritance is not sufficient, as setting `RequestedTheme` anywhere in between the source and target needs to "reset" this to default. I have a solution in, but it does not seem to work completely (switching to default theme from dark leaves font white), so I will need to work on that some more.

MartinZikmund committed:
chore: Fix ParentTheme test, fix TextBlock foreground theming

MartinZikmund committed:
test: Themed controls test

davidjohnoliver commented:
@MartinZikmund Nice stuff, I'll take a look. :)

For the commented-out test, I suspect the behaviour might've been different for the older UWP 1709, but I can't remember for sure. It's fine to restore it if it matches the behaviour of the UWP version we're currently targeting.

MartinZikmund commented:
> @MartinZikmund Nice stuff, I'll take a look. :)
> 
> For the commented-out test, I suspect the behaviour might've been different for the older UWP 1709, but I can't remember for sure. It's fine to restore it if it matches the behaviour of the UWP version we're currently targeting.

@davidjohnoliver Tested with the two latest SDKs and it seems both throws, so the exception should be ok.

davidjohnoliver commented on src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.cs:
```diff
@@ -393,6 +393,71 @@ Brush Foreground
+						ForegroundProperty,
+						foregroundSource.RequestedTheme == ElementTheme.Light ?
+							SolidColorBrushHelper.Black : SolidColorBrushHelper.White,
+						DependencyPropertyValuePrecedences.ImplicitStyle);
```
I'm not sure I understand the reasoning with the precedences. Why are you using `ImplicitStyle` here?

MartinZikmund replied:
My reasoning here is that I need to make sure that RequestedTheme has precedence over the Inherited value - which ImplicitStyle does. The reason is that setting RequestedTheme somewhere in the tree "stops" the Foreground property inheritance

davidjohnoliver replied:
Can you add a comment here explaining this?

I'm also curious how this would interact with issue of style vs inheritance precedence - https://github.com/unoplatform/uno/issues/3443

MartinZikmund replied:
Hmm, good question 🤔 . I am now realizing, I will probably also need a similar "trick" added to the `Control` foreground logic

MartinZikmund replied:
I was wrong, this is `TextBlock`-specific, it behaves a bit differently. I have updated the code to be less invasive about the implicit setting and all test I have done seem ok now. Let's see what UI test suite says. Regarding #3443, these changes, unfortunately, don't fix it, but it still behaves the same at least :-D . There is still the outstanding issue with visual states however (that theme is not properly applied when a button is pressed, etc.). If the tests pass, we could probably merge this and let that as a separate issue, as for existing scenarios there is no change.

davidjohnoliver commented on src/Uno.UI/UI/Xaml/ResourceDictionary.cs:
```diff
@@ -89,32 +89,34 @@ public void Add(object key, object value)
-		public bool TryGetValue(object key, out object value, bool shouldCheckSystem)
+		public bool TryGetValue(object key, out object value) => TryGetValue(key, out value, shouldCheckSystem: true, ElementTheme.Default);
+
+		/// <remarks>Should become internal in the future?</remarks>
```
oops! Yes this should be internal, same for the `ContainsKey()` overload

MartinZikmund replied:
Should I change it within this PR?

davidjohnoliver replied:
Yes, go ahead

MartinZikmund replied:
Done :-)

davidjohnoliver commented on src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:
```diff
@@ -317,6 +323,12 @@ public void SetBinding(object target, string dependencyProperty, BindingBase bin
 
+		public void SetResourceBinding(DependencyProperty dependencyProperty, object resourceKey, bool isTheme, object context)
+		{
+			var binding = new ResourceBinding(resourceKey, isTheme, context, _precedenceOverride ?? DependencyPropertyValuePrecedences.Local);
```
ah, nice

MartinZikmund committed:
chore: Addressing comments

MartinZikmund committed:
chore: Requested theme foreground override

davidjohnoliver commented on src/Uno.UI/UI/Xaml/Application.cs:
```diff
@@ -185,35 +185,9 @@ internal void SetRequestedTheme(ApplicationTheme requestedTheme)
-				}
-			}
+			FrameworkElement.PropagateThemeScope(
+				Windows.UI.Xaml.Window.Current.Content,
```
pretty edge-case scenario, but what happens if by some chance RequestedTheme has been set on `Windows.UI.Xaml.Window.Current.Content`?

MartinZikmund replied:
That is actually covered - the second argument for the `PropagateThemeScope` method just says what is the requested theme of the outside scope, if any. Inside the `UpdateActualTheme` method is called, which checks the processed element's `RequestedTheme` and gives it precedence.

MartinZikmund commented:
Oh ok, maybe there are still issues 😂😂😂
![image](https://user-images.githubusercontent.com/1075116/87958133-7473f500-cab1-11ea-8294-69a3cf541f30.png)

MartinZikmund committed:
chore: Enable exception on missing resource dictionary keys

MartinZikmund committed:
chore: Cleanup

MartinZikmund committed:
chore: Move theme propagation to loading

MartinZikmund commented:
@davidjohnoliver Will we be able to merge this in some form so the element theme support can be worked on further?

davidjohnoliver commented:
@MartinZikmund We should be able to merge this, providing it doesn't introduce any regressions.

'No regressions' probably means preserving the current special-casing for `Window.Current.Root`, for which changing the `RequestedTheme` just changes the whole app theme behind the scenes, at least until we're sure that the current element-level logic fully matches it.

jeromelaban commented:
Is there something that changed about this PR, particularly the last comment ?

MartinZikmund commented:
> Is there something that changed about this PR, particularly the last comment ?

@jeromelaban Ah, sorry, no. I didn't get to it since TreeView/TabView. Once I finish prepping TabView for merging, I will continue with this one to rebase on master and make sure to have it as a "alternative" for now, not to break the global setting.
